Allow CIDR notation in query string query

wbowling opened the pull request:
Allow IP ranges to be specified with CIDR notation in a query string query.

I refactored the cidrMaskToMinMax method from IPv4RangeBuilder so it could be used for the IpFieldMapper as well.

Closes #7464

rjernst commented on core/src/main/java/org/elasticsearch/common/network/InetAddresses.java:
```diff
@@ -23,10 +25,12 @@
 public class InetAddresses {
     private static int IPV4_PART_COUNT = 4;
     private static int IPV6_PART_COUNT = 8;
+    private static final Pattern MASK_PATTERN = Pattern.compile("[\\.|/]");
```
I think this should be an actual pattern for a ipv4 address? Otherwise the code below would allow passing eg `0/0/0/0/0` and parse it as an ip?

wbowling replied:
Sure, that was just the one way the range builder was doing it. I'll change it to use a Matcher and a stricter pattern.

jpountz commented on core/src/main/java/org/elasticsearch/common/network/InetAddresses.java:
```diff
@@ -16,6 +16,8 @@
 
 package org.elasticsearch.common.network;
 
+import org.elasticsearch.search.aggregations.bucket.range.ipv4.InternalIPv4Range;
```
can you move MAX_IP to this class and remove the above import?

jpountz commented on core/src/main/java/org/elasticsearch/index/mapper/ip/IpFieldMapper.java:
```diff
@@ -206,6 +209,23 @@ public BytesRef indexedValueForSearch(Object value) {
+                }
+                if (fromTo != null) {
+                    return rangeQuery(fromTo[0] < 0 ? Double.NEGATIVE_INFINITY : fromTo[0],
+                            fromTo[1] < 0 ? Double.POSITIVE_INFINITY : fromTo[1], true, false);
```
can you use `null` instead of POSITIVE/NEGATIVE_INFINITY? this is the way the rangeQuery method expects to be notified about open bounds

jpountz commented:
Left two minor comments but this looks good to me. I like how this changes the termQuery method so that all query parsers will benefit from this change (`query_string`, `simple_query_string` and `match` in particular).

rmuir commented:
> I like how this changes the termQuery method so that all query parsers will benefit from this change (query_string, simple_query_string and match in particular).

Note that "/" is a reserved character in the lucene query language. In particular it looks for content between two unescaped "/"'s and treats that as a regular expression.

jasontedor commented on core/src/main/java/org/elasticsearch/common/network/InetAddresses.java:
```diff
@@ -23,10 +25,13 @@
 public class InetAddresses {
     private static int IPV4_PART_COUNT = 4;
     private static int IPV6_PART_COUNT = 8;
+    private static final Pattern MASK_PATTERN = Pattern.compile("(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})/(\\d{1,3})");
```
I don't think that this is right since it will match 256.512.256.512 but that is not a valid IP address.

jpountz replied:
I'm tempted to also do it in a follow-up pr as this code is really just moved around.

jasontedor replied:
Also captured in #14862.

jasontedor commented on core/src/main/java/org/elasticsearch/common/network/InetAddresses.java:
```diff
@@ -354,4 +359,59 @@ private static InetAddress bytesToInetAddress(byte[] addr) {
+        if (!matcher.matches()) {
+            return null;
+        }
+        int addr = (( Integer.parseInt(matcher.group(1)) << 24 ) & 0xFF000000)
```
This will silently ignore octets that are captured by the regular expression, but are in fact not valid octets.

jpountz replied:
Agreed but let's handle this in a different PR as this is a pre-existing issue.

jasontedor replied:
I don't think that we should move this code into the InetAddresses.java source file. That code is from the Guava code base, is licensed to the Guava authors (see the license header). By moving this code which is not from Guava here we will create a confusing situation with respect to the licensing of the code.

jasontedor replied:
> Agreed but let's handle this in a different PR as this is a pre-existing issue (the method was just moved).

@jpountz Opened #14862.

wbowling commented:
Ok moved the code to IpFieldMapper so it's out of the Guava code, moved MAX_IP and changed to null instead of POSITIVE/NEGATIVE_INFINITY for the term query.

Some parts have been moved twice now so let me know if I should squash this all into one commit

jpountz commented:
Thanks  @wbowling. I merged it manually via 18c9eba40a044df0e4601e2438239f714a93a9e5

cbuescher commented:
@jpountz initial commit message mentions this closes #7464, apparently this didn't work automatically. Should I close there as well or am In missing something?

jpountz commented:
Indeed! I just closed the issue.
